pkg/kv/kvserver/gc/gc.go

```diff
@@ -381,149 +443,509 @@ func processReplicatedKeyRange(
+//     range
+//   - batchGCKeysBytesThreshold - maximum byte size of all deleted key versions
+//     per request
+//   - maxPendingKeysSize - maximum number of bytes taken by pending requests in
+//     memory
+//
+// Batcher tries to find the longest sequence of keys to delete and if the
+// sequence is not found (because of non collectable data inbetween) it will
+// resort to sending request with point deletions.
+// To achieve this behaviour batcher will collect point batches in parallel to
+// tracking the sequence length. Once minimum sequence length is reached, it
+// stops collecting point batches since we are guaranteed to send a clear range.
+// If however it reaches maxPendingKeysSize limit before clearRangeMinKeys is
+// met (that would happen if keys are particularly large) then the oldest batch
+// is sent, and clear range end is moved to the lowest key if sent batch.
+// Once non-garbage key is found, if sequence length is at or above
+// clearRangeMinKeys threshold, GC will first send relevant portion of first
+// points batch (up to the first key covered by clear range) and then clear
+// range request for the remainder of keys. After that it drops all pending
+// batches.
+// If however the sequence is too short then batcher will send all previous
+// batches up to the current one. Record current key as an end of clear range
+// span (remember that we are iterating from back to front), record current
+// batch counters in partialPointBatchCounters and restart sequence
+// search. partialPointBatchCounters is required to make accurate counter
+// updates when clear range starts mid points batch. It covers part of the
+// first batch from its beginning to the point where clear range could start,
+// and remainder of the batch is covered by counters for clear range span.
+// Described behaviour support the following pending batches invariant:
+//   - only the oldest batch in the history could be split by non-garbage key in
+//     the middle, other batches always cover contiguous sets of keys.
+type gcKeyBatcher struct {
+	gcKeyBatcherThresholds
+
+	// prevWasNewest indicates that we are at oldest version of the key.
+	// its value is a saved isNewest from previous key.
+	prevWasNewest bool
+	// Points batches are accumulated until we cross clear range threshold.
+	// There's always at least one points batch initialized.
+	pointsBatches []pointsBatch
+	// totalMemUsed sum of memUsed of all points batches cached to avoid
+	// recomputing on every check.
+	totalMemUsed int
+
+	// Tracking of clear range requests, only used if clearRangeEnabled is true.
+
+	// clearRangeEndKey contains a key following first garbage key found. It would
+	// be greater or equal to the last key (lowest, collected last) of the first
+	// (collected first, containing highest keys) points batch. It is guaranteed
+	// that there are no garbage keys between this key and the start of first
```

Comment:
Can this be equal to the last key? Doesn't it have to be greater than it in order to cover it?

Also, this should be "no keys" rather than "no garbage keys", right? Because any live keys would be removed.

Reply:
Yeah, that is correct. It could be last when we start iterating, but we will always tighten it to the batch[0][0].Next() before sending.